test/state/host.cpp

```diff
@@ -0,0 +1,369 @@
+    sender_acc.balance -= value;
+    new_acc.balance += value;  // The new account may be prefunded.
+
+    auto create_msg = msg;
+    create_msg.input_data = nullptr;
+    create_msg.input_size = 0;
+
+    // Execution can modify the state, iterators are invalidated.
+    auto result = m_vm.execute(*this, m_rev, create_msg, msg.input_data, msg.input_size);
+    if (result.status_code != EVMC_SUCCESS)
+    {
+        result.create_address = msg.recipient;
+        return result;
+    }
+
+    auto gas_left = result.gas_left;
+    assert(gas_left >= 0);
+
+    bytes_view code{result.output_data, result.output_size};
+    if (m_rev >= EVMC_SPURIOUS_DRAGON && code.size() > 0x6000)
+        return evmc::Result{EVMC_OUT_OF_GAS};
+
+    const auto cost = static_cast<int64_t>(code.size()) * 200;
+    gas_left -= cost;
+    if (gas_left < 0)
+    {
+        evmc::Result r{EVMC_OUT_OF_GAS};
+
+        if (m_rev == EVMC_FRONTIER)
+        {
+            r.status_code = EVMC_SUCCESS;
+            r.gas_left = result.gas_left;
+        }
+
+        return r;
+    }
+
+    // Reject EF code.
+    if (m_rev >= EVMC_LONDON && !code.empty() && code[0] == 0xEF)
+        return evmc::Result{EVMC_OUT_OF_GAS};
+
+    // TODO: The new_acc pointer is invalid because of the state revert implementation,
+    //       but this should change if state journal is implemented.
+    m_state.get(msg.recipient).code = code;
+
+    return evmc::Result{result.status_code, gas_left, result.gas_refund, msg.recipient};
+}
+
+evmc::Result Host::execute_message(const evmc_message& msg) noexcept
+{
+    if (msg.kind == EVMC_CREATE || msg.kind == EVMC_CREATE2)
+        return create(msg);
+
+    auto* const code_acc = m_state.get_or_null(msg.code_address);
+
+    if (msg.kind == EVMC_CALL)
+    {
+        // TODO: Should not create empty touched account?
+        assert(evmc::address{msg.recipient} == msg.code_address);
+        auto& recipient_acc = code_acc != nullptr ? *code_acc : m_state.create(msg.recipient);
+        recipient_acc.touched = true;
+
+        // Transfer value.
+        const auto value = intx::be::load<intx::uint256>(msg.value);
+        assert(m_state.get(msg.sender).balance >= value);
+        m_state.get(msg.sender).balance -= value;
+        recipient_acc.balance += value;
+    }
+
+    // Copy of the code. Revert will invalidate the account.
+    const auto code = code_acc != nullptr ? code_acc->code : bytes{};
+    return m_vm.execute(*this, m_rev, msg, code.data(), code.size());
+}
+
+evmc::Result Host::call(const evmc_message& orig_msg) noexcept
+{
+    const auto msg = prepare_message(orig_msg);
+    if (!msg.has_value())
+        return evmc::Result{EVMC_OUT_OF_GAS, orig_msg.gas};  // Light exception.
+
+    auto state_snapshot = m_state;
+    auto destructs_snapshot = m_destructs.size();
+    auto access_addresses_snapshot = m_accessed_addresses;
+    auto logs_snapshot = m_logs.size();
+
+    auto result = execute_message(*msg);
```

Comment:
can be const?

Reply:
No because we have to move it.

Reply:
Do you mean to return value? copy will be elided even if it's const I think

Reply:
Yes, but `Result` is move-only type